shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs

```diff
@@ -1,57 +1,52 @@
 import qualified Cardano.Ledger.Crypto as CryptoClass
-import Cardano.Ledger.Era
-import Cardano.Ledger.Torsor (Torsor (..))
-import Cardano.Ledger.Val (DecodeNonNegative, Val)
+import Cardano.Ledger.Era (Era (Crypto))
+import Cardano.Ledger.Shelley.Constraints (TxBodyConstraints)
 import Shelley.Spec.Ledger.Coin (Coin)
-import Shelley.Spec.Ledger.Hashing (EraIndependentTxBody, HashAnnotated (..))
-
---------------------------------------------------------------------------------
--- Shelley Era
---------------------------------------------------------------------------------
+import Shelley.Spec.Ledger.Keys (hashWithSerialiser)
+import Shelley.Spec.Ledger.MetaData (MetaData (MetaData), MetaDataHash (MetaDataHash), ValidateMetadata (hashMetadata, validateMetadata), validMetaDatum)
+import Shelley.Spec.Ledger.Scripts (MultiSig)
+import Shelley.Spec.Ledger.Tx (TxBody, ValidateScript (hashScript, validateScript), hashMultiSigScript, validateNativeMultiSigScript)
 
 data ShelleyEra c
 
 instance CryptoClass.Crypto c => Era (ShelleyEra c) where
   type Crypto (ShelleyEra c) = c
 
-type instance Value (ShelleyEra c) = Coin
-
-type TxBodyConstraints era =
-  ( ChainData (TxBody era),
-    AnnotatedData (TxBody era),
-    HashAnnotated (TxBody era) era,
-    HashIndex (TxBody era) ~ EraIndependentTxBody
-  )
-
-type ShelleyBased era =
-  ( Era era,
-    -- Value constraints
-    Val (Value era),
-    Compactible (Value era),
-    DecodeNonNegative (Value era),
-    ChainData (Value era),
-    Eq (CompactForm (Value era)),
-    SerialisableData (Value era),
-    SerialisableData (CompactForm (Value era)),
-    ChainData (Delta (Value era)),
-    SerialisableData (Delta (Value era)),
-    Torsor (Value era),
-    -- TxBody constraints
-    TxBodyConstraints era,
-    -- Script constraints
-    ChainData (Script era),
-    AnnotatedData (Script era),
-    -- Metadata constraints
-    ChainData (Metadata era),
-    AnnotatedData (Metadata era)
-  )
+--------------------------------------------------------------------------------
+-- Core instances
+--------------------------------------------------------------------------------
+
+type instance Core.Value (ShelleyEra _c) = Coin
+
+type instance Core.TxBody (ShelleyEra c) = TxBody (ShelleyEra c)
+
+type instance Core.Script (ShelleyEra c) = MultiSig (ShelleyEra c)
+
+type instance Core.Metadata (ShelleyEra c) = MetaData
+
+--------------------------------------------------------------------------------
+-- Ledger data instances
+--------------------------------------------------------------------------------
+
+instance
+  (CryptoClass.Crypto c, TxBodyConstraints (ShelleyEra c)) =>
+  ValidateScript (ShelleyEra c)
+  where
+  validateScript = validateNativeMultiSigScript
+  hashScript = hashMultiSigScript
+
+instance CryptoClass.Crypto c => ValidateMetadata (ShelleyEra c) where
+  hashMetadata = MetaDataHash . hashWithSerialiser toCBOR
+  validateMetadata (MetaData m) = all validMetaDatum m
```

Comment:
Probably unrelated to this PR, but I saw a `deepseq` here in some other implementation. Is that intentional?

Reply:
That was in the Shelley-MA package, related to the new optional scripts. The idea there is that we want to verify that scripts included in the metadata are valid. Scripts are valid by construction, so this check comes down to making sure they're fully deserialised.